tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py

```diff
@@ -452,78 +453,115 @@ def set_compare_custom_predict_fn(self, predict_fn):
     return self
 
   def _convert_json_to_tf_examples(self, examples):
+    self._set_uses_json_input(True)
     tf_examples = []
     for json_ex in examples:
       ex = tf.train.Example()
-      for feat in json_ex:
-        if isinstance(json_ex[feat], (int, long)):
-          ex.features.feature[feat].int64_list.value.append(json_ex[feat])
-        elif isinstance(json_ex[feat], Number):
-          ex.features.feature[feat].float_list.value.append(json_ex[feat])
-        else:
-          ex.features.feature[feat].bytes_list.value.append(
-            json_ex[feat].encode('utf-8'))
-      tf_examples.append(ex)
+      # JSON examples can be lists of values (for xgboost models for instance),
+      # or dicts of key/value pairs.
+      if isinstance(json_ex, list):
+        self._set_uses_json_list(True)
+        feature_names = self.config.get('feature_names')
+        for (i, value) in enumerate(json_ex):
+          # If feature names have been provided, use those feature names instead
+          # of list indices for feature name when storing as tf.Example.
+          if feature_names and len(feature_names) > i:
+            feat = feature_names[i]
+          else:
+            feat = str(i)
+          if isinstance(json_ex[i], (int, long)):
```

Comment:
nit, you can pull this code chunk into a function add_single_feature(key, value, ex) to avoid repetition.

Reply:
done